### ISSUE TITLE
getAvailableLocaleCodes not refreshing when new locales are fetched dynamically

### ISSUE THREAD

**Issue by boris-petrov** [fullcalendar/fullcalendar#4799]
In our code we fetch new locales dynamically and then eval them. This seems to work fine, that is, the calendar uses the new locale just fine when we do `calendar.setOption('locale', locale)`. However, `calendar.getAvailableLocaleCodes()` returns only the ones that were available when the calendar was created and doesn't reflect the newly added ones. In version 3 we did the same with `$.fullCalendar.locales` and it did refresh when we evaled new locales.

P.S. Another thing is that the `firstDay` doesn't change when `calendar.setOption('locale', locale)` is invoked. It only dynamically changes when a new calendar is created. Should I open a new issue for that?

**Comment by acerix** [comment]
`getAvailableLocaleCodes` is not mentioned in the docs so I don't beleive it's intended to be used like that.  Would it work to keep the list of locales outside the calendar?

`firstDay` should update when the locale changes as it does in this demo:

https://fullcalendar.io/docs/locale-demo

If it's not working, please submit an issue here with a [runnable, stripped-down demonstration of the bug](http://fullcalendar.io/wiki/Reduced-Test-Cases/).

**Comment by boris-petrov** [comment]
Hi, thanks for the support.

You're right, `getAvailableLocaleCodes` is not documented, I saw it in the code and figured out that I was just not seeing it in the documentation. I need it because I don't know whether I have to pass to `fullcalendar` a two-letter or a four-letter locale. So when I eval the locale files with that function I can check which of the two is the correct one. But I have to create a `new Calendar` in order for it to "refresh".

`firstDay` is the same - if I eval a new locale and do a `calendar.setOption('locale', locale)`, the calendar doesn't change the `firstDay`. I have to create a `new Calendar`. Note that this is only true if I `eval` a new locale - that is, it was not available when I construct the first calendar.

Please tell me if you don't understand. I can try creating a reproduction. But I think it's pretty clear - adding a locale dynamically doesn't refresh some of the data in an already existing calendar.

**Comment by acerix** [comment]
The locales need to be loaded before the calendar is initialized, is there a reason you can't load them in advance? It might be easier to include all the locales with `locales-all.js` if you are using lots of them.

**Comment by boris-petrov** [comment]
@acerix - well, that's the point, we don't know how many are going to be used (if any at all) so that's why we want to save tens/hundreds of kilobytes of needless data being transferred and executed. For that we just download and import the locales dynamically as needed.

Note that this used to work in `fullcalendar` 3 and it is a regression. If you believe it is fine (after all, it is a major version bump), you can close the issue. But I think it is a nice feature to have and perhaps I'm not the only one doing something like that.

**Comment by acerix** [comment]
Loading locales after the calendar was initialized wasn't documented in v3 so I think it's not really a regression but I can leave this open as a feature request.

**Comment by arshaw** [comment]
Would you be able to post a [runnable, stripped-down demonstration of the bug](https://fullcalendar.io/reduced-test-cases)? Would really appreciate it because the time saved reproducing will be time spent fixing. I'm closing this issue until then.

**Comment by boris-petrov** [comment]
@arshaw - well, this issue is even worse in FullCalendar version 5. The locales there are not browserified and one cannot simply import them dynamically in a website:

```typescript
var enGb = {
  code: "en-gb",
  week: {
    dow: 1, // Monday is the first day of the week.
    doy: 4  // The week that contains Jan 4th is the first week of the year.
  }
};

export default enGb;
```

I have configured my bundler to copy these locales to my distribution and then I have setup code to dynamically load locales as requested by the user - in order to not download to that user all locales upfront.

In version 4 that worked fine because the locale files were browserified. But then `getAvailableLocaleCodes` didn't "refresh" when a new locale was added - I had to create a new Calendar instance for it to work.

In version 5 I have to manually "fix" the locale files to work in a browser. And I removed my workarounds with creating a new Calendar instance and all that.

So the issue can be closed, yes, even though the underlying problem is still there albeit not relevant any more to me.

Thanks for the support!
